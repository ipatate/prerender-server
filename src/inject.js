// @flow
import cheerio from 'cheerio';

/*
 * @param source String
 * inject async attribute in script tags
 */
export const addAsync = (source: string): string => {
  const $ = cheerio.load(source);
  $('script').each((e, v) => {
    if ($(v).attr('src') !== undefined) {
      $(v).attr('async', 'async');
    }
  });
  return $.html();
};
