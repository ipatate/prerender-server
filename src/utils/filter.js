// @flow
/**
 * @description format url for regex
 * @param {RegExp} regex
 * @param {string} replaceBy
 */
export const escapeString = (
  regex: RegExp | string = '',
  replaceBy: string = '\\$1',
): Function => (str: string): string => str.replace(regex, replaceBy);

/**
 * @description find format in string
 * @param {RegExp} regex
 */
export const findInString = (regex: RegExp | string = ''): Function => (
  str: string,
): boolean => str.search(regex) > -1;
