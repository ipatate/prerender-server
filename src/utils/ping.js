// @flow
import ping from 'ping';

export default (url: string): Promise<boolean> =>
  new Promise(resolve => ping.sys.probe(url, r => resolve(r)));
