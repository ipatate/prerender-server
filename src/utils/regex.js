// @flow
export const urlToRegex = (regex: RegExp = /([.?*+^$[\]\\(){}|-])/): RegExp =>
  new RegExp(regex, 'g'); // eslint-disable-line

// clean url for file name
export const cleanName = (regex: RegExp = /[./:-?{}()#*+\\|]/): RegExp =>
  new RegExp(regex, 'g'); // eslint-disable-line

// extension file
export const extensionRegex = (
  regex: RegExp = /\.(xml|less|png|jpg|jpeg|gif|pdf|doc|txt|ico|rss|zip|mp3|rar|exe|wmv|doc|avi|ppt|mpg|mpeg|tif|wav|mov|psd|ai|xls|mp4|m4a|swf|dat|dmg|iso|flv|m4v|torrent|ttf|woff|woff2)$/, // prettier-ignore
): RegExp => new RegExp(regex);

// external request
export const tiersRegex = (
  regex: RegExp = /(googletagmanager|google-analytics|gstatic|googleapis|facebook|twitter|zopim)/,
): RegExp => new RegExp(regex);
