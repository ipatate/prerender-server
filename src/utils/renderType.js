// @flow

export const renderType = (
  renderTypeOptions: Array<string> = ['html', 'jpeg', 'png', 'pdf'],
): Function => (renderType: string = ''): string => {
  let type = 'html';
  if (renderTypeOptions.indexOf(renderType) > -1) {
    type = renderType;
  }
  return type;
};
