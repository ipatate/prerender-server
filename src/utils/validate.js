// @flow
import isUrl from 'is-url';

export const validateUrl = (url: string): boolean => isUrl(url);
