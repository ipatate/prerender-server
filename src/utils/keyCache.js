// @flow
import {URL} from 'url';
import {cleanName} from './regex';
import {escapeString} from './filter';

export const getUtc = (): string =>
  new Date()
    .toJSON()
    .slice(0, 19)
    .replace(/:/g, '-');

export const getKeyCache = (
  urlPage: string,
  extension: string = '.html',
): string => {
  // escape for file name
  const escapeForFileName = escapeString(cleanName(), '-');
  let urlScrap;
  try {
    urlScrap = new URL(urlPage);
  } catch (e) {
    return `no-valid-url/${getUtc()}`;
  }
  // dir name
  const dirName = escapeForFileName(urlScrap.hostname);
  // format url for file name
  const fileName = escapeForFileName(
    `${dirName}${urlScrap.pathname}${urlScrap.search}`,
  );
  return `${dirName}/${fileName}${extension}`;
};
