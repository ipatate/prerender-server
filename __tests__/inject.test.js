import {addAsync} from '../src/utils/inject';

test('inject async in script tag', () => {
  const html = `<html><head><script charset="utf-8" src="/b09d3e62070d304b57fb.Home.js"></script></head><body><main><div></div></main></div><script type="text/javascript">window.__INITIAL_LANG__ = {"fr":{"translation":{"Home":"Accueil","history":"Histoire"}}};</script><script type="text/javascript">window.__INITIAL_STATE__ = {"ux":{"showLoader":false,"languages":["fr"],"currentLanguage":"fr","errorMessage":""},"films":[]};</script><script src="/main.dc74ab992de17e7b4922.js" async></script></body></html>`;

  const result = addAsync(html);
  expect(result).toMatchSnapshot();
});
