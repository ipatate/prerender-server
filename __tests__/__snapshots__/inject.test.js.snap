// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inject async in script tag 1`] = `"<html><head><script charset=\\"utf-8\\" src=\\"/b09d3e62070d304b57fb.Home.js\\" async=\\"async\\"></script></head><body><main><div></div></main><script type=\\"text/javascript\\">window.__INITIAL_LANG__ = {\\"fr\\":{\\"translation\\":{\\"Home\\":\\"Accueil\\",\\"history\\":\\"Histoire\\"}}};</script><script type=\\"text/javascript\\">window.__INITIAL_STATE__ = {\\"ux\\":{\\"showLoader\\":false,\\"languages\\":[\\"fr\\"],\\"currentLanguage\\":\\"fr\\",\\"errorMessage\\":\\"\\"},\\"films\\":[]};</script><script src=\\"/main.dc74ab992de17e7b4922.js\\" async=\\"async\\"></script></body></html>"`;
